docs/source/composer_model.rst

```diff
@@ -0,0 +1,248 @@
+   trainer.fit()
+
+Both the |forward| and |loss| methods are provided the ``batch`` directly
+from the dataloader. We leave the unpacking of that batch into inputs and targets
+to the user, since it can very depending on the task.
+
+We also provide several common classes for various tasks,
+specifically:
+
+-  :class:`.ComposerClassifier` - classification tasks with a cross entropy
+   loss and accuracy metric.
+-  :class:`.TIMM` - creates classification models from the popular `TIMM`_
+   library.
+-  :class:`.ComposerTransformer` - base class for 🤗 `Transformers`_ models.
+
+..note ::
+
+    Users from other frameworks such as pytorch lightning, may be used to
+    defining a ``training_step`` method, which groups the forward and loss
+    together. However, many of our algorithmic methods (such as
+    label smoothing or selective backprop) need to intercept and modify the
+    loss. For this reason, we split into two separate methods.
+
+By convention, we define our pytorch layers in the ``self.model``
+attribute of :class:`.ComposerModel`, which we encourage because it's easier
+to extract the underlying model for inference when training is
+completed. However, this is not enforced, and users can configure the
+layers directly in the class if they prefer.
+
+Metrics
+-------
+
+To compute metrics during training, implement the methods:
+
+.. code:: python
+
+   def validate (self, batch) -> outputs, targets:
+       ...
+
+   def metrics(self, train=False) -> Metrics:
+       ...
+
+where ``Metrics`` should be compatible with the torchmetrics package. We
+require that the output of :meth:`.ComposerModel.validate` be consumable by
+torchmetrics. Specifically, in our trainer loop, the validation loop
+does something like this:
+
+.. code:: python
+
+   metrics = model.metrics(train=False)
+
+   for batch in val_dataloader:
+       outputs, targets = model.validate(batch)
+       metrics.update(outputs, targets)  # implements the torchmetrics interface
+
+   metrics.compute()
+
+A full example of a validation implementation would be:
+
+.. code:: python
+
+   class ComposerClassifier(ComposerModel):
+
+       def __init__(self):
+           super().__init__()
+           self.model = torchvision.models.resnet18()
+           self.train_accuracy = torchmetrics.Accuracy()
+           self.val_accuracy = torchmetrics.Accuracy()
+
+         ...
+
+       def validate(self, batch):
+
+           inputs, targets = batch
+           outputs = self.model(inputs)
+           return outputs, targets
+
+       def metrics(self, train=False):
+           # defines which metrics to use in each phase of training
+           return self.train_accuracy if train else self.val_accuracy
+
+.. note::
+
+    No need to set ``model.eval()`` or ``torch.no_grad()`` — we take care
+    of that in our trainer. ``torchmetrics`` also handles metrics logging
+    when using distributed training.
+
+
+Logging Results
```

Comment:
Do we need this detail here? This seems much more related to the trainer than the model...